diff --git a/lib/parser_block.mjs b/lib/parser_block.mjs
index 36e0109..33ce0fd 100644
--- a/lib/parser_block.mjs
+++ b/lib/parser_block.mjs
@@ -14,6 +14,7 @@ import r_blockquote from './rules_block/blockquote.mjs'
 import r_hr from './rules_block/hr.mjs'
 import r_list from './rules_block/list.mjs'
 import r_reference from './rules_block/reference.mjs'
+import r_abbr_def from './rules_block/abbr_def.mjs'
 import r_html_block from './rules_block/html_block.mjs'
 import r_heading from './rules_block/heading.mjs'
 import r_lheading from './rules_block/lheading.mjs'
@@ -28,6 +29,7 @@ const _rules = [
   ['blockquote', r_blockquote, ['paragraph', 'reference', 'blockquote', 'list']],
   ['hr', r_hr, ['paragraph', 'reference', 'blockquote', 'list']],
   ['list', r_list, ['paragraph', 'reference', 'blockquote']],
+  ['abbr_def', r_abbr_def, ['paragraph', 'reference']],
   ['reference', r_reference],
   ['html_block', r_html_block, ['paragraph', 'reference', 'blockquote']],
   ['heading', r_heading, ['paragraph', 'reference', 'blockquote']],
diff --git a/lib/parser_core.mjs b/lib/parser_core.mjs
index 4337554..7623e8f 100644
--- a/lib/parser_core.mjs
+++ b/lib/parser_core.mjs
@@ -12,6 +12,7 @@ import r_normalize from './rules_core/normalize.mjs'
 import r_block from './rules_core/block.mjs'
 import r_inline from './rules_core/inline.mjs'
 import r_linkify from './rules_core/linkify.mjs'
+import r_abbr_replace from './rules_core/abbr_replace.mjs'
 import r_replacements from './rules_core/replacements.mjs'
 import r_smartquotes from './rules_core/smartquotes.mjs'
 import r_text_join from './rules_core/text_join.mjs'
@@ -21,6 +22,7 @@ const _rules = [
   ['block', r_block],
   ['inline', r_inline],
   ['linkify', r_linkify],
+  ['abbr_replace', r_abbr_replace],
   ['replacements', r_replacements],
   ['smartquotes', r_smartquotes],
   // `text_join` finds `text_special` tokens (for escape sequences)
diff --git a/lib/rules_block/abbr_def.mjs b/lib/rules_block/abbr_def.mjs
new file mode 100644
index 0000000..a538f91
--- /dev/null
+++ b/lib/rules_block/abbr_def.mjs
@@ -0,0 +1,48 @@
+// Process abbreviation definitions, e.g. `*[HTML]: HyperText Markup Language`
+//
+export default function abbr_def (state, startLine, endLine, silent) {
+  let labelEnd
+  let pos = state.bMarks[startLine] + state.tShift[startLine]
+  const max = state.eMarks[startLine]
+
+  if (pos + 2 >= max) { return false }
+
+  if (state.src.charCodeAt(pos++) !== 0x2A/* * */) { return false }
+  if (state.src.charCodeAt(pos++) !== 0x5B/* [ */) { return false }
+
+  const labelStart = pos
+
+  for (; pos < max; pos++) {
+    const ch = state.src.charCodeAt(pos)
+    if (ch === 0x5B /* [ */) {
+      return false
+    } else if (ch === 0x5D /* ] */) {
+      labelEnd = pos
+      break
+    } else if (ch === 0x5C /* \ */) {
+      pos++
+    }
+  }
+
+  if (labelEnd < 0 || state.src.charCodeAt(labelEnd + 1) !== 0x3A/* : */) {
+    return false
+  }
+
+  // abbreviations need an env to store definitions in
+  if (!state.env) { return false }
+
+  if (silent) { return true }
+
+  const label = state.src.slice(labelStart, labelEnd).replace(/\\(.)/g, '$1')
+  const title = state.src.slice(labelEnd + 2, max).trim()
+  if (label.length === 0) { return false }
+  if (title.length === 0) { return false }
+  if (!state.env.abbreviations) { state.env.abbreviations = {} }
+  // prepend ':' to avoid conflict with Object.prototype members
+  if (typeof state.env.abbreviations[':' + label] === 'undefined') {
+    state.env.abbreviations[':' + label] = title
+  }
+
+  state.line = startLine + 1
+  return true
+}
diff --git a/lib/rules_core/abbr_replace.mjs b/lib/rules_core/abbr_replace.mjs
new file mode 100644
index 0000000..1125e1e
--- /dev/null
+++ b/lib/rules_core/abbr_replace.mjs
@@ -0,0 +1,88 @@
+// Enclose defined abbreviations in <abbr> tags across inline text tokens.
+//
+import { escapeRE, arrayReplaceAt, lib } from '../common/utils.mjs'
+
+// ASCII characters in Cc, Sc, Sm, Sk categories we should terminate on.
+const OTHER_CHARS = ' \r\n$+<=>^`|~'
+
+const UNICODE_PUNCT_RE = lib.ucmicro.P.source
+const UNICODE_SPACE_RE = lib.ucmicro.Z.source
+
+export default function abbr_replace (state) {
+  const blockTokens = state.tokens
+
+  if (!state.env || !state.env.abbreviations) { return }
+
+  const keys = function () {
+    return Object.keys(state.env.abbreviations).map(function (x) {
+      return x.substr(1)
+    }).sort(function (a, b) {
+      return b.length - a.length
+    }).map(escapeRE)
+  }
+
+  const regSimple = new RegExp('(?:' + keys().join('|') + ')')
+
+  const regText = '(^|' + UNICODE_PUNCT_RE + '|' + UNICODE_SPACE_RE +
+                  '|[' + OTHER_CHARS.split('').map(escapeRE).join('') + '])' +
+          '(' + keys().join('|') + ')' +
+          '($|' + UNICODE_PUNCT_RE + '|' + UNICODE_SPACE_RE +
+                  '|[' + OTHER_CHARS.split('').map(escapeRE).join('') + '])'
+
+  const reg = new RegExp(regText, 'g')
+
+  for (let j = 0, l = blockTokens.length; j < l; j++) {
+    if (blockTokens[j].type !== 'inline') { continue }
+    let tokens = blockTokens[j].children
+
+    // We scan from the end, to keep position when new tags added.
+    for (let i = tokens.length - 1; i >= 0; i--) {
+      const currentToken = tokens[i]
+      if (currentToken.type !== 'text') { continue }
+
+      let pos = 0
+      const text = currentToken.content
+      reg.lastIndex = 0
+      const nodes = []
+
+      // fast regexp run to determine whether there are any abbreviated words
+      // in the current token
+      if (!regSimple.test(text)) { continue }
+
+      let m
+
+      while ((m = reg.exec(text))) {
+        if (m.index > 0 || m[1].length > 0) {
+          const token = new state.Token('text', '', 0)
+          token.content = text.slice(pos, m.index + m[1].length)
+          nodes.push(token)
+        }
+
+        const token_o = new state.Token('abbr_open', 'abbr', 1)
+        token_o.attrs = [['title', state.env.abbreviations[':' + m[2]]]]
+        nodes.push(token_o)
+
+        const token_t = new state.Token('text', '', 0)
+        token_t.content = m[2]
+        nodes.push(token_t)
+
+        const token_c = new state.Token('abbr_close', 'abbr', -1)
+        nodes.push(token_c)
+
+        reg.lastIndex -= m[3].length
+        pos = reg.lastIndex
+      }
+
+      if (!nodes.length) { continue }
+
+      if (pos < text.length) {
+        const token = new state.Token('text', '', 0)
+        token.content = text.slice(pos)
+        nodes.push(token)
+      }
+
+      // replace current node
+      blockTokens[j].children = tokens = arrayReplaceAt(tokens, i, nodes)
+    }
+  }
+}
